pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java

```diff
@@ -60,140 +62,166 @@
+                    .topic(topic);
+            if (producerName != null) {
+                builder.producerName(producerName);
+            }
+
+            return builder
+                    .property("application", "pulsarfunction")
+                    .property("fqfn", fqfn).create();
         }
 
-        @Override
-        public void sendOutputMessage(TypedMessageBuilder<T> msg, Record<T> record) throws Exception {
-            msg.sendAsync();
+        protected Producer<T> getProducer(String destinationTopic) {
+            return getProducer(destinationTopic, null, destinationTopic);
         }
 
-        @Override
-        public void close() throws Exception {
-            if (null != producer) {
+        protected Producer<T> getProducer(String producerId, String producerName, String topicName) {
+
+            Producer<T> producer = publishProducers.get(producerId);
+
+            if (producer == null) {
                 try {
-                    producer.close();
+                    Producer<T> newProducer = createProducer(
+                            client,
+                            topicName,
+                            producerName,
+                            schema,
+                            fqfn);
+
+                    Producer<T> existingProducer = publishProducers.putIfAbsent(producerId, newProducer);
+
+                    if (existingProducer != null) {
+                        // The value in the map was not updated after the concurrent put
+                        newProducer.close();
+                        producer = existingProducer;
+                    } else {
+                        producer = newProducer;
+                    }
+
                 } catch (PulsarClientException e) {
-                    log.warn("Fail to close producer for processor {}", pulsarSinkConfig.getTopic(), e);
+                    log.error("Failed to create Producer while doing user publish", e);
+                    throw new RuntimeException(e);
                 }
             }
+            return producer;
         }
-    }
-
-    private class PulsarSinkAtLeastOnceProcessor implements PulsarSinkProcessor<T> {
-        private Producer<T> producer;
 
         @Override
-        public void initializeOutputProducer(String outputTopic, Schema<T> schema, String fqfn) throws Exception {
-            this.producer = AbstractOneOuputTopicProducers.createProducer(
-                    client, pulsarSinkConfig.getTopic(), null, schema, fqfn);
+        public void close() throws Exception {
+            List<CompletableFuture<Void>> closeFutures = new ArrayList<>(publishProducers.size());
+            for (Map.Entry<String, Producer<T>> entry: publishProducers.entrySet()) {
+                String topicId = entry.getKey();
+                Producer<T> producer = entry.getValue();
+                closeFutures.add(producer.closeAsync().exceptionally(throwable -> {
```

Comment:
should we just do: `closeFutures.add(producer.closeAsync());` because logging will not help much;